File: ios/Sources/RS.swift
// RS.swift — Reed-Solomon over GF(256), straight port of rs.py (encode + full decode:
// syndromes, Berlekamp-Massey, Chien search, magnitudes by GF Gaussian elimination).
// Systematic RS(n, n-nsym), corrects nsym/2 byte errors per block.
import Foundation

enum RS {
    static let tables: (exp: [Int], log: [Int]) = {
        var exp = [Int](repeating: 0, count: 512), log = [Int](repeating: 0, count: 256)
        var x = 1
        for i in 0 ..< 255 {
            exp[i] = x; log[x] = i
            x <<= 1
            if x & 0x100 != 0 { x ^= 0x11d }
        }
        for i in 255 ..< 510 { exp[i] = exp[i - 255] }
        return (exp, log)
    }()

    static func mul(_ a: Int, _ b: Int) -> Int {
        a == 0 || b == 0 ? 0 : tables.exp[tables.log[a] + tables.log[b]]
    }

    static func genPoly(_ nsym: Int) -> [Int] {
        var g = [1]
        for i in 0 ..< nsym {
            var g2 = [Int](repeating: 0, count: g.count + 1)
            for j in 0 ..< g.count { g2[j] ^= g[j] }                       // g * x
            for j in 0 ..< g.count { g2[j + 1] ^= mul(g[j], tables.exp[i]) }
            g = g2
        }
        return g
    }

    static func encode(_ data: [UInt8], _ nsym: Int) -> [UInt8] {   // one block -> +parity
        let g = Array(genPoly(nsym).dropFirst())
        var par = [Int](repeating: 0, count: nsym)
        for d in data {
            let f = Int(d) ^ par[0]
            par.removeFirst(); par.append(0)
            if f != 0 { for j in 0 ..< nsym { par[j] ^= mul(f, g[j]) } }
        }
        return data + par.map { UInt8($0) }
    }

    static func decode(_ codeIn: [UInt8], _ nsym: Int) -> (data: [UInt8], ok: Bool) {
        let n = codeIn.count
        var code = codeIn.map { Int($0) }
        let (EXP, LOG) = tables
        func syndromes() -> [Int] {
            (0 ..< nsym).map { j in
                var s = 0
                for i in 0 ..< n where code[i] != 0 {
                    s ^= EXP[(LOG[code[i]] + j * (n - 1 - i)) % 255]
                }
                return s
            }
        }
        let syn = syndromes()
        if !syn.contains(where: { $0 != 0 }) {
            return (Array(codeIn[0 ..< n - nsym]), true)
        }
        var C = [1], B = [1], L = 0, m = 1, b = 1                    // Berlekamp-Massey
        for i in 0 ..< nsym {
            var d = syn[i]
            let jmax = min(L, C.count - 1)
            if jmax >= 1 {
                for j in 1 ... jmax where j <= i && C[j] != 0 && syn[i - j] != 0 {
                    d ^= EXP[LOG[C[j]] + LOG[syn[i - j]]]
                }
            }
            if d == 0 { m += 1; continue }
            let coef = ((LOG[d] - LOG[b]) % 255 + 255) % 255
            let Bp = [Int](repeating: 0, count: m) + B.map { $0 != 0 ? EXP[LOG[$0] + coef] : 0 }
            var Cn = C + [Int](repeating: 0, count: max(0, Bp.count - C.count))
            for j in 0 ..< Bp.count { Cn[j] ^= Bp[j] }
            if 2 * L <= i { L = i + 1 - L; B = C; b = d; m = 1 } else { m += 1 }
            C = Cn
        }
        if L > nsym / 2 { return (Array(codeIn[0 ..< n - nsym]), false) }
        var powers = [Int]()                                          // Chien search
        for e in 0 ..< n {
            var v = 0
            for (j, cf) in C.enumerated() where cf != 0 {
                v ^= EXP[(LOG[cf] + ((255 - e * j % 255) % 255)) % 255]
            }
            if v == 0 { powers.append(e) }
        }
        if powers.count != L { return (Array(codeIn[0 ..< n - nsym]), false) }
        // magnitudes: solve V e = s, V[j][k] = alpha^(p_k * j), GF Gaussian elimination
        var A = (0 ..< L).map { j in powers.map { EXP[($0 * j) % 255] } + [syn[j]] }
        for col in 0 ..< L {
            guard let piv = (col ..< L).first(where: { A[$0][col] != 0 }) else {
                return (Array(codeIn[0 ..< n - nsym]), false)
            }
            A.swapAt(col, piv)
            let inv = EXP[255 - LOG[A[col][col]]]
            for j in 0 ... L { A[col][j] = mul(A[col][j], inv) }
            for r in 0 ..< L where r != col && A[r][col] != 0 {
                let f = A[r][col]
                for j in 0 ... L { A[r][j] ^= mul(f, A[col][j]) }
            }
        }
        for (k, p) in powers.enumerated() { code[n - 1 - p] ^= A[k][L] }
        for j in 0 ..< nsym {                                         // verify the fix
            var s = 0
            for i in 0 ..< n where code[i] != 0 {
                s ^= EXP[(LOG[code[i]] + j * (n - 1 - i)) % 255]
            }
            if s != 0 { return (Array(codeIn[0 ..< n - nsym]), false) }
        }
        return (code[0 ..< n - nsym].map { UInt8($0) }, true)
    }

    static func selfTest() -> Bool {              // encode -> corrupt nsym/2 -> decode
        var rng = SystemRandomNumberGenerator()
        for nsym in [32, 64] {
            let data = (0 ..< 255 - nsym).map { _ in UInt8.random(in: 0 ... 255, using: &rng) }
            var code = encode(data, nsym)
            var idx = Set<Int>()
            while idx.count < nsym / 2 { idx.insert(Int.random(in: 0 ..< 255, using: &rng)) }
            for i in idx { code[i] ^= UInt8.random(in: 1 ... 255, using: &rng) }
            let (dec, ok) = decode(code, nsym)
            if !ok || dec != data { return false }
        }
        return true
    }
}

File: transmitter/rs.py
# rs.py — Reed-Solomon over GF(256), numpy-vectorized across equal-size blocks.
# Systematic RS(n, n-nsym): corrects up to nsym//2 byte errors per block.
import numpy as np

_EXP = np.zeros(512, np.int32)
_LOG = np.zeros(256, np.int32)
x = 1
for i in range(255):
    _EXP[i] = x; _LOG[x] = i
    x <<= 1
    if x & 0x100: x ^= 0x11d
_EXP[255:510] = _EXP[:255]

def _mul(a, b):            # elementwise GF mul (arrays)
    out = _EXP[_LOG[a] + _LOG[b]]
    return np.where((a == 0) | (b == 0), 0, out)

def _gen_poly(nsym):
    g = np.array([1], np.int32)
    for i in range(nsym):
        g2 = np.zeros(len(g) + 1, np.int32)
        g2[:-1] ^= g                        # g * x
        g2[1:] ^= _mul(g, _EXP[i])          # + g * alpha^i
        g = g2
    return g

def encode(data, nsym):    # data (nblk, k) -> (nblk, k+nsym)
    g = _gen_poly(nsym)[1:]
    nblk, k = data.shape
    par = np.zeros((nblk, nsym), np.int32)
    for j in range(k):
        f = data[:, j] ^ par[:, 0]
        par = np.roll(par, -1, 1); par[:, -1] = 0
        par ^= _mul(f[:, None], g[None, :])
    return np.concatenate([data, par], 1).astype(np.uint8)

_SYN_T = {}                # (n, nsym) -> log-power matrix, built once

def _syndromes(code, nsym):
    # S_j = C(alpha^j) for all blocks/syndromes at once. This used to be a Horner loop
    # over the 255 columns (255 python iterations of _mul per decode) — profiled at
    # ~40ms/frame on the Surface, paid even by CLEAN frames, since syndromes are always
    # computed. Direct evaluation instead: S_j = XOR_i c_i * alpha^(j*(n-1-i)). The
    # log of each power IS (j*(n-1-i)) mod 255 — no table lookup to build it — so the
    # whole thing is one fancy-index EXP[LOG[c] + T], a zero-mask (log of 0 is
    # undefined; GF mul by 0 is 0), and an XOR-reduce. ~19x64x255 int32 = 1.2MB, ~2ms.
    nblk, n = code.shape
    if (n, nsym) not in _SYN_T:
        _SYN_T[(n, nsym)] = (np.arange(nsym)[:, None] * (n - 1 - np.arange(n))[None, :]) % 255
    prod = _EXP[_LOG[code][:, None, :] + _SYN_T[(n, nsym)][None, :, :]]
    return np.bitwise_xor.reduce(np.where(code[:, None, :] == 0, 0, prod), 2)

def _solve(V, s):          # GF Gaussian elimination: V e = s, V (L,L)
    L = V.shape[0]
    A = np.concatenate([V, s[:, None]], 1).astype(np.int32)
    for col in range(L):
        piv = next((r for r in range(col, L) if A[r, col]), None)
        if piv is None: return None
        A[[col, piv]] = A[[piv, col]]
        A[col] = _mul(A[col], _EXP[255 - _LOG[A[col, col]]])
        for r in range(L):
            if r != col and A[r, col]:
                f = A[r, col].copy()
                A[r] ^= _mul(f, A[col])
    return A[:, -1]

def decode(code, nsym):    # (nblk, n) -> (data, ok mask)
    code = code.astype(np.int32)
    nblk, n = code.shape
    syn = _syndromes(code, nsym)
    ok = ~syn.any(1)
    fixed = []
    EXPL, LOGL = _EXP.tolist(), _LOG.tolist()   # python ints: ~10x faster in the BM loop
    for bi in np.where(~ok)[0]:
        s = syn[bi].tolist()
        C, B, L, m, b = [1], [1], 0, 1, 1       # Berlekamp-Massey
        for i in range(nsym):
            d = s[i]
            for j in range(1, min(L, len(C) - 1) + 1):
                cj, sij = C[j], s[i - j]
                if cj and sij: d ^= EXPL[LOGL[cj] + LOGL[sij]]
            if d == 0:
                m += 1; continue
            coef = (LOGL[d] - LOGL[b]) % 255
            Bp = [0] * m + [EXPL[LOGL[x] + coef] if x else 0 for x in B]
            Cn = C + [0] * max(0, len(Bp) - len(C))
            for j, x in enumerate(Bp): Cn[j] ^= x
            if 2 * L <= i:
                L, B, b, m = i + 1 - L, C[:], d, 1
            else:
                m += 1
            C = Cn
        if L > nsym // 2: continue
        C = np.array(C, np.int32)
        e_arr = np.arange(n)             # Chien: sigma(alpha^-e) == 0 -> error at power e
        val = np.zeros(n, np.int32)
        for j, cf in enumerate(C):
            if cf: val ^= _mul(int(cf), _EXP[(-e_arr * j) % 255])
        powers = e_arr[val == 0]
        if len(powers) != L: continue
        # magnitudes: solve S_j = sum_k e_k * (alpha^{p_k})^j  for j = 0..L-1
        V = np.array([[_EXP[(p * j) % 255] for p in powers] for j in range(L)], np.int32)
        ev = _solve(V, np.array(s[:L], np.int32))
        if ev is None: continue
        code[bi, n - 1 - powers] ^= ev
        fixed.append(bi)
    if fixed:                            # verify corrected blocks in one vectorized pass
        chk = _syndromes(code[fixed], nsym)
        ok[np.array(fixed)[~chk.any(1)]] = True
    return code[:, :n - nsym].astype(np.uint8), ok

if __name__ == "__main__":
    rng = np.random.default_rng(0)
    for nsym in (8, 16, 32):
        data = rng.integers(0, 256, (40, 200 - nsym)).astype(np.uint8)
        code = encode(data.astype(np.int32), nsym)
        nerr = nsym // 2
        corr = code.copy()
        for bi in range(40):
            idx = rng.choice(200, nerr, replace=False)
            corr[bi, idx] ^= rng.integers(1, 256, nerr).astype(np.uint8)
        dec, ok = decode(corr, nsym)
        print(f"nsym={nsym}: ok={ok.all()} exact={np.array_equal(dec, data)}")
